test/jdk/java/util/stream/BuiltInGatherersTest.java

```diff
@@ -294,43 +309,51 @@ public void testMapConcurrentAPIandContract(Config config) throws InterruptedExc
 
         // Test cancellation during short-circuiting
         if (config.streamSize > 2) {
-            final var firstLatch = new CountDownLatch(1);
-            final var secondLatch = new CountDownLatch(1);
-            final var cancellationLatch = new CountDownLatch(config.streamSize - 2); // all but two will get cancelled
+            final var tasksToCancel = config.streamSize - 2;
+            final var firstReady = new CountDownLatch(1);
+            final var lastDone = new CountDownLatch(1);
+            final var tasksCancelled = new CountDownLatch(tasksToCancel);
+
+            final var tasksWaiting = new Semaphore(0);
 
             final var result =
                 config.stream()
-                      .gather(
+                    .gather(
                             Gatherers.mapConcurrent(config.streamSize(), i -> {
                                 switch (i) {
                                     case 1 -> {
-                                        try {
-                                            firstLatch.await(); // the first waits for the last element to start
-                                        } catch (InterruptedException ie) {
-                                            throw new IllegalStateException(ie);
-                                        }
+                                        firstReady.countDown();
+                                        awaitSensibly(lastDone);
                                     }
 
-                                    case Integer n when n == config.streamSize - 1 -> { // last element
-                                        firstLatch.countDown(); // ensure that the first element can now proceed
+                                    case Integer n when n == config.streamSize - 1 -> {
+                                        awaitSensibly(firstReady);
+                                        while(tasksWaiting.getQueueLength() < tasksToCancel) {
```

Comment:
@AlanBateman This is the only part of this I really don't like. Any better suggestion as to "wait for N things waiting" in a test? 🤔

Reply:
> @AlanBateman This is the only part of this I really don't like. Any better suggestion as to "wait for N things waiting" in a test? 🤔

Semaphore::getQueueLength is more for monitoring purposes so the usual does seem a bit unusual here. There are a number of ways this could be done, maybe the simplest is for tasksWaiting to be an AtomicInteger and have the default just increment it and sleep-for-a-day.

Reply:
@AlanBateman Unfortunately that doesn't really work as there's then a window between incrementing the counter and going to sleep. So while it might execute correctly in 99.9999% of cases I don't want to have a jbs bug filed which will be one of those things where you can't recall it 6-12 months down the line 😓

Reply:
> @AlanBateman Unfortunately that doesn't really work as there's then a window between incrementing the counter and going to sleep. So while it might execute correctly in 99.9999% of cases I don't want to have a jbs bug filed which will be one of those things where you can't recall it 6-12 months down the line 😓

Thread.sleep throws if called with the interrupt status set so a cancel in the window between before the sleep should be okay. I think in this case, it's whatever is simplest is reliable and easy to read/maintain.